streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1909,6 +1909,16 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
       }
     }
 
+  /**
+   * Runs the stream returning the first element of the stream. Later elements will not be consumed / have effects run
+   */
+  def runHead: ZIO[R, E, Option[A]] = take(1).runCollect.map(_.headOption)
```

Comment:
This implementation collects entire stream into a list. It can run out of memory. 
Should it be `ZSink.foldLeft(None){ case (a, _) => a }` ?

Reply:
The entire stream is at most 1 element after `take(1)` so this cannot run out of memory and does work infinite streams